File: 392. Is Subsequence/Solution.py
# Given two strings s and t, 
# return true if s is a subsequence of t, or false otherwise.
# i.e., "ace" is a subsequence of "abcde" while "aec" is not).

def isSubsequence(s: str, t: str) -> bool:
    s_pointer = 0
    t_pointer = 0

    while s_pointer < len(s) and t_pointer < len(t):
        # se o caractere atual de t coincide com o de s avança em s
        if s[s_pointer] == t[t_pointer]:
            s_pointer += 1
        # avança sempre no ponteiro de t
        t_pointer += 1

    # se percorrer todos os caracteres de s, significa que e uma subsequencia
    return s_pointer == len(s)


w1='ab'
w2='bdb'

print(isSubsequence(w1,w2))

File: 392. Is Subsequence/Solution.java
public class Solution {
    
    public static void main(String[] args) {

        String w1 = "xtt";
        String w2 = "acbe";
        
        System.out.println(isSubsequence(w1, w2));

    }

    public static boolean isSubsequence(String s, String t) {
        int t_pointer = 0;
        int s_pointer = 0;

        while (s_pointer < s.length() && t_pointer < t.length()) {
            if (s.charAt(s_pointer) == t.charAt(t_pointer)) {
                s_pointer += 1;
            }
            t_pointer += 1;
        }

        return s_pointer == s.length();
    }
}
